Firestore/core/src/firebase/firestore/util/path.h

```diff
@@ -20,37 +20,129 @@
 #include <string>
 #include <utility>
 
+#include "Firestore/core/src/firebase/firestore/util/string_apple.h"
 #include "absl/strings/string_view.h"
 
 namespace firebase {
 namespace firestore {
 namespace util {
 
-struct Path {
+/**
+ * An immutable native pathname string. Paths can be absolute or relative. Paths
+ * internally maintain their filesystem-native encoding.
+ *
+ * The intent of the API is that high-level code generally just uses UTF-8-
+ * encoded string-literals or strings for the segments and constructs a
+ * filesystem-native Path using APIs like Path::JoinUtf8.
+ *
+ * Lower level code may need to construct derived Paths from values obtained
+ * from the filesystem that already are in the right encoding.
+ */
+class Path {
+ public:
+#if defined(_WIN32)
+  using char_type = wchar_t;
+  using string_type = std::wstring;
+
+  static constexpr char_type kPreferredSeparator = L'\\';
+#else
+  using char_type = char;
+  using string_type = std::string;
+
+  static constexpr char_type kPreferredSeparator = '/';
+#endif  // defined(_WIN32)
+
+  static constexpr size_t npos = static_cast<size_t>(-1);
+
+  /**
+   * Creates a new Path from a UTF-8-encoded pathname.
+   */
+  static Path FromUtf8(absl::string_view utf8_pathname);
+
+#if defined(_WIN32)
+  /**
+   * Creates a new Path from a UTF-16-encoded pathname.
+   */
+  // absl::wstring_view does not exist :-(.
+  static Path FromUtf16(const wchar_t* begin, size_t size);
+#endif
+
+#if defined(__OBJC__)
+  /**
+   * Creates a new Path from the given NSString pathname.
+   */
+  static Path FromNSString(NSString* path) {
+    return FromUtf8(MakeStringView(path));
+  }
+#endif
+
+  Path() {
+  }
+
+  const string_type& native_value() const {
+    return pathname_;
+  }
+
+  const char_type* c_str() const {
+    return pathname_.c_str();
+  }
+
+  size_t size() const {
+    return pathname_.size();
+  }
+
+#if defined(_WIN32)
```

Comment:
Very optional: I understand that these functions are grouped logically. I would _personally_ find it easier to grasp if instead they were grouped by platform, which would minimize the number of `ifdef`s and allow to quickly see platform-specific interface. But it's a matter of preference, so if you find existing version more readable, or don't have a preference either way, then please just leave as is.